amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShutdownTaskTest.java

```diff
@@ -232,6 +234,63 @@ public final void testCallWhenLeaseLost() throws DependencyException, InvalidSta
         verify(leaseCoordinator, never()).getAssignments();
         verify(leaseRefresher, never()).createLeaseIfNotExists(Matchers.any(Lease.class));
         verify(leaseCoordinator, never()).dropLease(Matchers.any(Lease.class));
+        verify(leaseRefresher, never()).deleteLease(any(Lease.class));
+    }
+
+    /**
+     * Test method for {@link ShutdownTask#call()}.
+     * Tests that when both child shard leases are present, we are able to delete the parent shard.
+     */
+    @Test
+    public final void testParentShardLeaseDeletedAtShardEndHappyCase()
+            throws DependencyException, InvalidStateException, ProvisionedThroughputException {
+        shardInfo = new ShardInfo("shardId-0", concurrencyToken, Collections.emptySet(),
+                ExtendedSequenceNumber.LATEST);
+        task = new ShutdownTask(shardInfo, shardDetector, shardRecordProcessor, recordProcessorCheckpointer,
+                SHARD_END_SHUTDOWN_REASON, INITIAL_POSITION_TRIM_HORIZON, cleanupLeasesOfCompletedShards,
+                ignoreUnexpectedChildShards, leaseCoordinator, TASK_BACKOFF_TIME_MILLIS, recordsPublisher,
+                hierarchicalShardSyncer, NULL_METRICS_FACTORY, constructChildShards());
+
+        final List<String> childShards = Arrays.asList("childShard-0", "childShard-1");
+        final Lease lease0 = createLease(shardInfo.getLeaseKey(shardInfo, "childShard-0"), "leaseOwner", Collections.emptySet());
+        final Lease lease1 = createLease(shardInfo.getLeaseKey(shardInfo, "childShard-1"), "leaseOwner", Collections.emptySet());
+
+        when(recordProcessorCheckpointer.lastCheckpointValue()).thenReturn(ExtendedSequenceNumber.SHARD_END);
+        final Lease heldLease = createLease("shardId-0", "leaseOwner", Collections.singleton("parentShardId"), childShards);
+        when(leaseCoordinator.getCurrentlyHeldLease("shardId-0")).thenReturn(heldLease);
+        when(leaseCoordinator.leaseRefresher()).thenReturn(leaseRefresher);
+        when(leaseRefresher.getLease(ShardInfo.getLeaseKey(shardInfo, "childShard-0"))).thenReturn(lease0);
+        when(leaseRefresher.getLease(ShardInfo.getLeaseKey(shardInfo, "childShard-1"))).thenReturn(lease1);
+
+        final TaskResult result = task.call();
+        assertNull(result.getException());
+        verify(leaseRefresher, times(2)).createLeaseIfNotExists(Matchers.any(Lease.class));
+        verify(leaseRefresher, times(1)).deleteLease(heldLease);
+    }
+
+    /**
+     * Test method for {@link ShutdownTask#call()}.
+     * Tests that if the child shard leases are missing, we fail fast and don't delete the parent shard lease.
```

Comment:
Should we add a test case where only one (out of two) child shards is present in lease table to make sure we don't delete parent's lease?

Reply:
Yeah, will add this as a test case.